File: react_native_app/src/features/productos/domain/usecases/getProductos.ts
import { UseCase } from '../../../../core/usecase/usecase';
import { Producto } from '../entities/Producto';
import { ProductoRepository } from '../repositories/ProductoRepository';

/**
 * Obtiene la lista de productos. El parámetro indica si se debe forzar la
 * recarga desde la API (true) o usar el estado en memoria (false).
 */
export class GetProductos implements UseCase<Producto[], boolean> {
  constructor(private readonly repository: ProductoRepository) {}

  call(forceRefresh = false): Promise<Producto[]> {
    return this.repository.getProductos(forceRefresh);
  }
}

File: flutter_app/lib/features/productos/domain/usecases/get_productos.dart
import '../../../../core/usecase/usecase.dart';
import '../entities/producto.dart';
import '../repositories/producto_repository.dart';

/// Obtiene la lista de productos. El parámetro indica si se debe forzar la
/// recarga desde la API (true) o usar el estado en memoria (false).
class GetProductos implements UseCase<List<Producto>, bool> {
  final ProductoRepository repository;
  GetProductos(this.repository);

  @override
  Future<List<Producto>> call(bool forceRefresh) =>
      repository.getProductos(forceRefresh: forceRefresh);
}
